pygmt/tests/test_figure.py

```diff
@@ -92,6 +92,46 @@ def test_figure_savefig_exists():
         fname.unlink()
 
 
+def test_figure_savefig_geotiff():
+    """
+    Make sure .tif generates a normal TIFF file and .tiff generates a GeoTIFF
+    file.
+    """
+    fig = Figure()
+    fig.basemap(region=[0, 10, 0, 10], projection="M10c", frame=True)
+
+    # Save as GeoTIFF
+    geofname = Path("test_figure_savefig_geotiff.tiff")
+    fig.savefig(geofname)
+    assert geofname.exists()
+    # The .pgw should not exist
+    assert not geofname.with_suffix(".pgw").exists()
+
+    # Save as TIFF
+    fname = Path("test_figure_savefig_tiff.tif")
+    fig.savefig(fname)
+    assert fname.exists()
+
+    # Check if a TIFF is georeferenced or not
+    try:
+        # pylint: disable=import-outside-toplevel
+        import rioxarray
+        from rasterio.errors import NotGeoreferencedWarning
+
+        # GeoTIFF
+        with rioxarray.open_rasterio(geofname) as xds:
+            assert xds.rio.crs is not None
```

Comment:
The output CRS from this 10cm Mercator projection map is:

```
PROJCS["unknown",GEOGCS["unknown",DATUM["unknown",SPHEROID["unknown",6378137,298.257220143428]],PRIMEM["Greenwich",0],UNIT["degree",0.0174532925199433,AUTHORITY["EPSG","9122"]]],PROJECTION["Mercator_1SP"],PARAMETER["central_meridian",5],PARAMETER["scale_factor",1],PARAMETER["false_easting",0],PARAMETER["false_northing",0],UNIT["metre",1,AUTHORITY["EPSG","9001"]],AXIS["Easting",EAST],AXIS["Northing",NORTH]]
```

I tried dragging and dropping the `test_figure_savefig_geotiff.tiff` into QGIS, and the projection showed up as 'unknown' (same as rioxarray above), but it defaulted to plotting on longitude 0-10, latitude 0-10:

![image](https://github.com/GenericMappingTools/pygmt/assets/23487320/064a2a1a-ab81-43ef-bd2b-2074672419ed)

However, the blue gridlines (actual longitude/latitude lines from QGIS) don't match up with the GeoTIFF (from psconvert), see the top left corner around 10°N, where the black/white zebra grid markings don't line up with the blue lines.

![image](https://github.com/GenericMappingTools/pygmt/assets/23487320/670abc16-aca9-4ee4-b80e-4eedfc405fe4)

Not sure what the default projection is that is output from psconvert, but I think we should make sure there isn't a bug somewhere. Maybe we should try a few other projections besides `M10c` and see if the output is ok?

@EJFielding, could you share with us the projection (`-J`) parameter you used to generate the overlay plot over Mexico at https://github.com/GenericMappingTools/pygmt/issues/2658#issue-1879147130? Just wanted to know how you're aligning the grids properly.

Reply:
Oh wait nevermind, I think I had a cached `tiff.aux.xml` file from trying a few other projections like `-JEPSG:3857`, and it was rendering the GeoTIFF in the wrong projection. Deleting all the files and resaving them seems to work better now:

![image](https://github.com/GenericMappingTools/pygmt/assets/23487320/c596964d-e9d7-438c-b7a8-ab94eb1df096)

There's still a 900m gap between the 9°N blue longitude line and the black/white zebra grid border, but I think that's just because of the default dpi resolution being a bit coarse.

Reply:
Does the frame make trouble here? The **psconvert -W** (https://docs.generic-mapping-tools.org/latest/psconvert.html) says:

> Be aware, however, that different results are obtained depending on the image contents and if the [-B](https://docs.generic-mapping-tools.org/latest/psconvert.html#b) option has been used or not. The trouble with the [-B](https://docs.generic-mapping-tools.org/latest/psconvert.html#b) option is that it creates a frame and very likely its annotations. That introduces pixels outside the map data extent, and therefore the map extents estimation will be wrong. To avoid this problem use --MAP_FRAME_TYPE=inside option which plots all annotations and ticks inside the image and therefore does not compromise the coordinate computations. Pay attention also to the cases when the plot has any of the sides with whites only because than the algorithm will fail miserably as those whites will be eaten by the Ghostscript. In that case you really must use [-B](https://docs.generic-mapping-tools.org/latest/psconvert.html#b) or use a slightly off-white color.

Reply:
Assuming that the `-B` means `-B` from `grdimage/basemap/etc` and not `psconvert`... I tried:

```python
with pygmt.config(MAP_FRAME_TYPE="inside"):
        fig.basemap(region=[0, 10, 0, 10], projection="M10c", frame="afg")
```

And it does look a bit better on the latitude (y-axis) side. The blue line and black lines align well on the latitude grid.

![image](https://github.com/GenericMappingTools/pygmt/assets/23487320/dad5fb00-2226-4980-aecd-26b81f621c63)

But on the longitude (x-axis) side, there still seems to be some offset (~900m) with the blue line not aligning with the black line:

![image](https://github.com/GenericMappingTools/pygmt/assets/23487320/7669d88b-c524-4c0e-a9cc-43d2853d5fb0)

Reply:
Could you please try to produce a figure using `Figure.coast` or `Figure.grdimage` without the `frame` parameter? The frame still has a thickness, which may affect the boundingbox or something.

Reply:
Still having an offset without the frame:

```python
fig = Figure()
with pygmt.config(MAP_FRAME_TYPE="inside"):
    fig.coast(shorelines=True, region=[0, 10, 0, 10], projection="M10c")
# Save as GeoTIFF
geofname = Path("test_figure_savefig_geotiff.tiff")
fig.savefig(geofname)
```

Top-left corner

![image](https://github.com/GenericMappingTools/pygmt/assets/23487320/4992a295-98ef-4c8b-8d94-544eeab4b6eb)

Bottom-right corner

![image](https://github.com/GenericMappingTools/pygmt/assets/23487320/8335eca5-6ce9-4431-ae00-1e449b0c5332)

I measured the size of one of the gray pixels and it was ~900+m, same as the offset. So it might be a case of gridline vs pixel registration (xref #487)? Though I'd think the offset would be half a pixel instead (~450m).

Reply:
> So it might be a case of gridline vs pixel registration (xref #487)?

But this is not a grid, so can't be the gridline/pixel registration issue. Anyway, if something doesn't make sense, then it must be an upstream bug.

Reply:
I found that the GeoTIFF images I was making with GMT classic looked better with the inside frame. I don't remember if having the frame outside made a difference in the geolocation.

I am not sure what the conventions are for "world" files, but I know that GDAL has the reference location for the image as the top-left corner of the top-left pixel, unless you change the registration. This is the same as "pixel" registration in GMT.

Reply:
The other thing to realize for checking image geolocation is that the coastline database used in GMT is something like 40 years old, and not very accurate. I would not rely on the GMT coastline to check the geolocations.

Reply:
@weiji14 When you have time, it would be better to submit an upstream issue for more in-depth discussions about the possible misalignments. I believe we can do nothing on the PyGMT side.